libafl_qemu/src/executor.rs

```diff
@@ -50,31 +53,103 @@ pub struct QemuExecutor<'a, C, CM, ED, ET, H, I, OT, S, SM> {
+    if let Some(qemu) = Qemu::get() {
+        // QEMU is already initialized, we have to route the signal to QEMU's handler or
+        // consider it as a host (i.e. fuzzer) signal
+
+        if qemu.is_running() {
+            // QEMU is running, we must determine if we are coming from qemu's signal handler or not
+            log::debug!("Signal has been triggered while QEMU was running");
+
+            match qemu.signal_ctx() {
+                QemuSignalContext::OutOfQemuSignalHandler => {
+                    // we did not run QEMU's signal handler, run it not
+                    log::debug!("It's a simple signal, let QEMU handle it first");
+
+                    qemu.run_signal_handler(signal.into(), info, puc);
+
+                    // if we are there, we can safely to execution
+                    return;
+                }
+                QemuSignalContext::InQemuSignalHandlerHost => {
+                    // we are running in a nested signal handling
+                    // and the signal is a host QEMU signal
+
+                    let si_addr = { info.si_addr() as usize };
+                    log::error!("QEMU Host crash crashed at addr 0x{si_addr:x}... Bug in QEMU or Emulator modules? Exiting.\n");
+
+                    if let Some(cpu) = qemu.current_cpu() {
+                        eprint!("QEMU Context:\n{}", cpu.display_context());
+                    }
+                }
+                QemuSignalContext::InQemuSignalHandlerTarget => {
+                    // we are running in a nested signal handler and the signal is a target signal.
+                    // run qemu hooks then report the crash.
+
+                    log::debug!("QEMU Target signal received that should be handled by host. Most likely a target crash.");
+
+                    log::debug!("Running crash hooks.");
+                    run_target_crash_hooks::<ET, I, S>(signal.into());
+
+                    assert!(data.maybe_report_crash::<E, EM, I, OF, S, Z>(None));
+
+                    if let Some(cpu) = qemu.current_cpu() {
+                        eprint!("QEMU Context:\n{}", cpu.display_context());
+                    }
+                }
+            }
+        } else {
+            // qemu is not running, it is a bug in LibAFL
+            let si_addr = { info.si_addr() as usize };
+            log::error!("The fuzzer crashed at addr 0x{si_addr:x}... Bug in the fuzzer? Exiting.");
+
+            let bsod = minibsod::generate_minibsod_to_vec(signal, info, context.as_deref());
+
+            if let Ok(bsod) = bsod {
+                if let Ok(bsod_str) = str::from_utf8(&bsod) {
+                    log::error!("\n{}", bsod_str);
+                } else {
+                    log::error!("convert minibsod to string failed");
+                }
+            } else {
+                log::error!("generate_minibsod failed");
+            }
+        }
     }
 
-    libafl_qemu_handle_crash(signal as i32, info, puc);
```

Comment:
I see there's no call to libafl_qemu_handle_crash anymore.
but doesn't this mean you won't go back to qemu side?? (which seems necessary)